spec/scheduling-tasks.md

```diff
@@ -0,0 +1,380 @@
+
+: <dfn>The posted task task source</dfn>
+:: This [=task source=] is used for tasks scheduled through {{Scheduler/postTask()}}.
+
+
+### Scheduler Task Queues ### {#sec-def-scheduler-task-queues}
+
+A <dfn>scheduler task queue</dfn> is a [=struct=] with the following [=struct/items=]:
+
+: <dfn for="scheduler task queue">priority</dfn>
+:: A {{TaskPriority}}.
+: <dfn for="scheduler task queue">tasks</dfn>
+:: A [=set=] of [=scheduler tasks=].
+
+<div algorithm>
+  To <dfn lt="create a scheduler task queue|creating a scheduler task queue">create a scheduler task queue</dfn>
+  with {{TaskPriority}} |priority|:
+
+  1. Let |queue| be a new [=scheduler task queue=].
+  1. Set |queue|'s [=scheduler task queue/priority=] to |priority|.
+  1. Set |queue|'s [=scheduler task queue/tasks=] to a new empty [=set=].
+  1. Return |queue|.
+</div>
+
+A [=scheduler task queue=] |queue|'s <dfn for="scheduler task queue">first runnable task</dfn>
+is the first [=scheduler task=] in |queue|'s [=scheduler task queue/tasks=] that is
+<a for="task">runnable</a>.
+
+Processing Model {#sec-scheduling-tasks-processing-model}
+---------------------
+
+
+### Queueing and Removing Scheduler Tasks ### {#sec-queuing-scheduler-tasks}
+
+<div algorithm="queue a scheduler task">
+  To <dfn lt="queue a scheduler task|queuing a scheduler task">queue a scheduler task</dfn>
+  on a [=scheduler task queue=] |queue|, which performs a series of steps |steps|,
+  given a numeric |enqueue order|, a [=task source=] |source|, and a [=document=] |document|:
+
+  1. Let |task| be a new [=scheduler task=].
+  1. Set |task|'s [=scheduler task/enqueue order=] to |enqueue order|.
+  1. Set |task|'s <a attribute for="task">steps</a> to |steps|.
+  1. Set |task|'s <a attribute for="task">source</a> to |source|.
+  1. Set |task|'s <a attribute for="task">document</a> to |document|.
+  1. Set |task|'s <a attribute for="task">script evaluation environment settings
+     object set</a> to a new empty [=set=].
+  1. [=set/Append=] |task| to |queue|'s [=scheduler task queue/tasks=].
+  1. Return |task|.
+
+  Issue: We should consider refactoring the HTML spec to add a constructor for
+  <a for="/">task</a>. One problem is we need the new task to be a
+  [=scheduler task=] rather than a <a for="/">task</a>.
+</div>
+
+<div algorithm>
+  To <dfn for="scheduler task queue">remove</dfn> a [=scheduler task=] |task| from
+  [=scheduler task queue=] |queue|, [=set/remove=] |task| from |queue|'s
+  [=scheduler task queue/tasks=].
+</div>
+
+### Scheduling Tasks ### {#sec-scheduler-alg-scheduling-tasks}
+
+<div algorithm="schedule a postTask task">
+  To <dfn lt="schedule a postTask task|scheduling a postTask task">schedule a postTask task</dfn>
+  for {{Scheduler}} |scheduler| given a {{SchedulerPostTaskCallback}} |callback| and
+  {{SchedulerPostTaskOptions}} options, run the following steps:
+
+  1. Let |result| be [=a new promise=].
+  1. Let |signal| be |options|["{{SchedulerPostTaskOptions/signal}}"].
```

Comment:
If you follow the above advice about not using nullable options, then you'll need to change these a bit to check for whether the value `[=map/exists=]`. You can convert not-present into null here if you want, so that it's easier to pass it through the algorithms.